keras_nlp/tokenizers/byte_pair_tokenizer.py

```diff
@@ -0,0 +1,460 @@
+            default=self.max_bpe_rank,
+        )
+
+    def get_vocabulary(self) -> List[str]:
+        """Get the tokenizer vocabulary as a list of strings tokens."""
+        return self.vocabulary.keys()
+
+    def vocabulary_size(self) -> int:
+        """Get the size of the tokenizer vocabulary."""
+        return len(self.vocabulary)
+
+    def id_to_token(self, id: int) -> str:
+        """Convert an integer id to a string token."""
+        # This will be slow, but keep memory usage down compared to building a
+        # . Assuming the main use case is looking up a few special tokens
+        # early in the vocab, this should be fine.
+
+        keys = self.get_vocabulary()
+        for token in keys:
+            if self.vocabulary[token] == id:
+                return token
+        return None
+
+    def token_to_id(self, token: str) -> int:
+        """Convert a string token to an integer id."""
+        return self.vocabulary[token]
+
+    def get_config(self):
+        # TODO(chenmoneygithub): now this class cannot be saved in TF format.
+        config = super().get_config()
+        config.update(
+            {
+                # Ideally a vocabulary would be saved as a plain text asset in
+                # the saved model. We have no good way to support this
+                # currently, so we save the vocabulary in the config.
+                "vocabulary": self.vocabulary,
+                "merges": self.merges,
+                "sequence_length": self.sequence_length,
+            }
+        )
+        return config
+
+    def tokenize(self, inputs):
+        if not isinstance(inputs, (tf.Tensor, tf.RaggedTensor)):
+            inputs = tf.convert_to_tensor(inputs)
+
+        scalar_input = inputs.shape.rank == 0
+        if scalar_input:
+            inputs = tf.expand_dims(inputs, 0)
+
+        # As re2 does not support lookahead match, we are using an alternative
+        # to insert a special token "६" before leading space of non-space
+        # characters and after the trailing space, e.g., " keras" will be
+        # "६ keras".
+        inputs = tf.strings.regex_replace(
+            inputs, rf"( )([^\s{self.special_space}])", r"६\1\2"
+        )
+        inputs = tf.strings.regex_replace(
+            inputs, rf"(\s{self.special_space})$", r"\1६"
+        )
+        raw_tokens = tf_text.regex_split(inputs, self.pat1, self.pat1)
+        # Second pass splits out the last whilespace char or "६".
+        raw_tokens = tf_text.regex_split(raw_tokens, self.pat2, self.pat2)
+        if raw_tokens.shape.rank > 2:
+            raw_tokens = raw_tokens.merge_dims(1, 2)
+        raw_tokens = self._remove_whitespace_placeholder(raw_tokens)
+        token_row_splits = raw_tokens.row_splits
+        flatten_tokens = raw_tokens.flat_values
+
+        # Check cache.
+        cache_lookup = self.cache.lookup(flatten_tokens)
+        cache_mask = cache_lookup == ""
+
+        count_nonzero_mask = tf.math.count_nonzero(
```

Comment:
this feels hard to read. we are looking for places where the cache lookup is is `""` but the original input is not `""`? why would the original input be `""`?

Could we do something more like `tf.math.reduce_any(cache_lookup == "" & flatten_tokens != "")`